Support running K8sVerifier tests on kind

joestringer opened the pull request:
After applying these changes, it's possible to run the K8sVerifier tests by:

```
make kind
make kind-image
cd test
FOCUS=K8sVerifier make k8s-kind
```

joestringer committed:
contrib: Mount Cilium dev dir into kind

Mount the Cilium git repository onto the kind nodes so that we can more
easily use this kind configuration in order to run particular ginkgo
tests that rely currently on the vagrant filesystem paths.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Add 'make k8s-kind' test target

This make target integrates my primary workflow for running the ginkgo
tests against a kind environment.

Usage (from test/):

    $ FOCUS=K8sVerifier make k8s-kind

Signed-off-by: Joe Stringer <joe@cilium.io>

nbusseneau commented on test/Makefile:
```diff
@@ -30,6 +31,26 @@ run:
+		K8S_VERSION="$$(kubectl version -o json | jq -r '.serverVersion | "\(.major).\(.minor)"')" \
+		ginkgo --focus "$(FOCUS)" --tags integration_tests -v -- \
+			-cilium.testScope=k8s \
+			--cilium.provision=false \
```
I think this should be
```suggestion
			-cilium.provision=false \
```

joestringer commented on test/Makefile:
```diff
@@ -30,6 +31,26 @@ run:
 	ginkgo --focus "K8s" --tags integration_tests -v -- --cilium.provision=$(PROVISION) --cilium.registryCredentials=$(REGISTRY_CREDENTIALS)
 
+# Match kind-image target in parent directory
+k8s-kind: export DOCKER_REGISTRY=localhost:5000
+k8s-kind:
```
btw this is a pretty bad make target name, but I don't have good ideas of what else to name it. I only named it `k8s-kind` because it's the `k8s` target but for the kind platform. If anyone has better ideas, I can change it.

christarazi replied:
I don't think it's terrible FWIW and I don't have better ideas. As long as the target is clearly documented on what it's going to do or the commands it's running are clear, then IMO it's fine.

pchaigno commented on test/bpf/verifier-test.sh:
```diff
@@ -289,9 +289,14 @@ function handle_developers {
 function main {
 	handle_args "$@"
 	handle_developers
+	mount_bpf
```
Why is this happening for Kind but not when running our usual Jenkins jobs?

It feels like an odd place to mount the bpffs. Shouldn't it be mounted as part of the Kind provisioning?

joestringer replied:
My guess is that the regular Jenkins jobs are guaranteed to install Cilium and so can rely on Cilium to set up bpffs for us. Whereas when I try this locally, I deliberately skip provisioning and jump straight to the K8sVerifier test.

I vaguely recall that for some reason kind by default ends up with sysfs mounted at /sys/fs/bpf 🤷 

I'm open to ideas on how to mount it as part of kind provisioning, but I'm not aware of any ways to make such configuration. The configurations I added in these commits are just mounting specific paths into the kind containers from the host using kind configuration objects, but we can't use this approach for bpffs or we may end up causing different nodes to reuse the same underlying FS and interfere with each other. In theory I guess the setup script could somehow go into the individual kind nodes and mount bpffs, not sure quite how to do this in a simple way (other than maybe deploying some dummy daemonset to mount the bpffs or something...?) . In the end it seemed simpler to just have a "check and mount bpffs" step in this script, since we know it will run where we need it to run and with the appropriate privileges, and this is a prerequisite to run the test.

pchaigno replied:
> My guess is that the regular Jenkins jobs are guaranteed to install Cilium and so can rely on Cilium to set up bpffs for us. Whereas when I try this locally, I deliberately skip provisioning and jump straight to the K8sVerifier test.

AFAIK, if K8sVerifier runs first, Cilium is not installed, as K8sVerifier doesn't install it itself. Hence why I'm surprised :slightly_smiling_face: 

Could we mount it as part of Verifier.go? We already deploy the test pod there, so it might make sense to also prepare the node (i.e., mount bpffs) in that same function.

joestringer replied:
That works too. I changed the patch to follow this suggestion, PTAL.

pchaigno reviewed (commented):
Thanks! This is awesome!

joestringer committed:
test/K8sVerifier: Mount bpffs in BeforeAll

Ensure that bpffs is mounted on the node as part of the test setup, so
that e.g. kind setups will be guaranteed to be set up correctly to run
this test locally.

Suggested-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>
